### PR TITLE
Add parameters to Models allowing robot joints to be parameterized

### PR DESCRIPTION
**Pull request by joaomoura24**
Changes for adding option of optimizing a selection of the joints by specifying the joints that are parameterized when getting the robot model.

I tested all the changes for Inverse Kinematics, Inverse Differential Kinematics, and configuration planning.
I copied the figure_eight_plan.py file and made the necessary changes to keep one of the joints fixed, so the optimizer only optimizes 6 joints, here: [link](https://github.com/joaomoura24/optas/blob/add-sel_opt_joints/example/figure_eight_plan_6dof.py)

The naming of functions might not be ideal, so please provide changes to the naming or the way it is done, as long as it doesn't break the goal functionality.

But also bare in mind that this functionality will not be used a lot by all the users, so it's ok to be a bit more cumbersome.
The added changes do not break the normal working of optas for the typical case that you optimize all of the joints.

### CODE REVIEW

**Comment by joaomoura24** [comment]
Actually, lets hold the merge until beginning of January, cause just before leaving I found a bug which doesn't seem to come from my latest changes (it looks like it's from a new model I am using), but still I want to check.

File: optas/builder.py
Comment (cmower) [inline optas/builder.py:96]: The comments potentially need updating. However, I am thinking maybe move this code block to above the `# Setup decision variables` block. Then the `SXContainer`'s are already initialized. Then you can setup the decision variables and parameters as you have it without splitting up where the `SXContainer`'s are initialized.

File: optas/models.py
Comment (cmower) [inline optas/models.py:234]: Could you use the full name here, i.e. `parameter_joint_names`. For variables it is fine to abbreviate but for class properties and methods I prefer to keep them as descriptive as possible.

File: optas/models.py
Comment (cmower) [inline optas/models.py:238]: Same comment as above re naming convention.

File: optas/models.py
Comment (cmower) [inline optas/models.py:242]: Same comment as above re naming convention. Perhaps here use `optimized_joint_names`.

File: optas/models.py
Comment (cmower) [inline optas/models.py:246]: Same comment as above re naming convention.

File: optas/models.py
Comment (cmower) [inline optas/models.py:267]: Same comment as above re naming convention.

File: optas/models.py
Comment (cmower) [inline optas/models.py:271]: Same comment as above re naming convention.

File: optas/models.py
Comment (cmower) [inline optas/models.py:275]: Same comment as above re naming convention.

Comment (cmower) [inline optas/models.py:275]: Perhaps we should introduce acceleration limits also. Perhaps instead of multiple properties for each upper/lower limits we should have one method that returns the limits given a `time_deriv`.

File: optas/builder.py
Comment (cmower) [inline optas/builder.py:216]: Replace `vector` with `array`.

File: optas/builder.py
Comment (cmower) [inline optas/builder.py:236]: Replace `vector` with `array`.

File: optas/builder.py
Comment (cmower) [inline optas/builder.py:305]: Won't this break when `optimize_time=True`?

File: optas/builder.py
Comment (cmower) [inline optas/builder.py:299]: I'm not sure I see the utility for this method. Also, I think it may break for different instantiations of the `OptimizationBuilder` (see below). I guess you want to be able to do something like `q = builder.get_joint_states()`, where the joint state `q` is a mix of decision variables and parameters?

Since this method only applies for `RobotModel`'s perhaps `robot` should appear in the method name since it will never be used for `TaskModel`'s. Then the `model` (from `get_model`) should be checked to ensure the user is referencing a robot model rather than a task model (you can use `assert` for this), something like

```python
assert isinstance(model, RobotModel), "this method only applies to robot models"
```

File: optas/builder.py
Comment (cmower) [inline optas/builder.py:76]: This comment will need updating.

File: optas/builder.py
Comment (cmower) [inline optas/builder.py:88]: This will break when the `model` is a `TaskModel`. We need to think about how `TaskModel` handles parameters.

File: example/figure_eight_plan_6dof.py
Comment (cmower) [inline example/figure_eight_plan_6dof.py:35]: This will likely throw an error somewhere down the line. `param_joints` is expected to be a list, so this line should be
```python
param_joints = ['lwr_arm_0_joint']
```

**Review by cmower — CHANGES_REQUESTED** [review]
Thanks @joaomoura24, I like the look of this however there are some breaking points. The `OptimizationBuilder` needs to be modified to be able to handle `TaskModel`'s with parameters. There are also some conflicts with master due to some other changes already merged - you'll need to rebase this branch but I think that shouldn't be too much hassle (the changes are fairly small).

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:96]: Done.

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:234]: done!

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:238]: done!

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:242]: done.

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:246]: done.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py]: Don't know. I haven't tested with optimize_time=True.
How can we test?

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:267]: done.

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:275]: done.

Comment (joaomoura24) [reply optas/models.py:275]: Yes, but i would leave this for another issue/pull request. Otherwise it is already a lot we are changing here.

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:271]: done.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:216]: done.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:236]: done.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:299]: The method is needed (I am using it) for not breaking backwards compatibility when you are using get_joint_states, which should still work as expected.
I changed the name to get_robot_states_and_parameters().

Comment (joaomoura24) [reply optas/builder.py:299]: Added the asset.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:76]: I don't understand this comment.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:88]: Added if condition to use this only if RobotModel and do the previous thing if TaskModel.
Does that solve it?
I don't know how to test it for the case of TaskModel!

**Comment by cmower** [comment]
Thanks @joaomoura24, looks like there are conflicts with `models.py`. You'll need to rebase with the master branch and see what needs resolving. Thanks!

**Comment by joaomoura24** [comment]
@cmower, I rebased twice this branch.
It is know working, though i see that you made lots of changes in the functions naming.
Please, lets finalize this merge this changes before making other changes to the master branch.
I can meet you tomorrow to discuss this if needed.

File: example/figure_eight_plan_6dof.py
Comment (cmower) [inline example/figure_eight_plan_6dof.py:103]: No need to set parameters as zero, by default if you don't specify them then they are assumed to be zero. See

https://github.com/cmower/optas/blob/01efe251b13b2c615403e28d124da56679f17db0/optas/sx_container.py#L80-L86

File: example/figure_eight_plan_6dof.py
Comment (cmower) [inline example/figure_eight_plan_6dof.py:112]: This, for me, defeats the point of having parameters. The solution should automatically contain the parametrized joints as well as the optimized joint values since the optimized and parametrized joints are, by definition, part of the same model. This merging should happen automatically in the [`solve`](https://github.com/cmower/optas/blob/01efe251b13b2c615403e28d124da56679f17db0/optas/solver.py#L98-L100) method.

File: example/figure_eight_plan_6dof.py
Comment (cmower) [inline example/figure_eight_plan_6dof.py:103]: The number of parametrized joints should also be a property of the `RobotModel` class. Later also, this should be abstracted to the `Model` class.

**Review by cmower — CHANGES_REQUESTED** [review]
This is moving in the right direction. However, I am not keen how the user is required to piece together the model optimized variables and parameters - the information on the ordering in the model array is all contained within the `RobotModel` class so this merging should be automatic.

This functionality should be abstracted to the `Model` class so `TaskModel` can take advantage of it too. I mention this in the comments that this we can leave as functionality only implemented, for now, in the `RobotModel` class. Please could you create an issue so we can implement this abstraction later. 

Please see comments for specific changes needed.

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:269]: It get's easily confused with q.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:278]: Man, you were the one who asked me to rename this to robot instead of model as i had in the beginning.
This is going in circles. Which one do you prefer?

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:98]: Impossible to have the time to implement (meaning testing as well) this in TaskModel, and for now there is no clear purpose for that general feature.
We should focus on features that we need.
I can create an issue.

File: example/figure_eight_plan_6dof.py
Comment (joaomoura24) [reply example/figure_eight_plan_6dof.py:46]: This is an example of usage. It just happened that the current configuration in this example was zero. But had it been something else it would be necessary to specify the initial configuration which might change for each call of the optimizer (after build).

File: example/figure_eight_plan_6dof.py
Comment (joaomoura24) [reply example/figure_eight_plan_6dof.py:112]: This was done this way to avoid major changes that would break the main use case where we optimize all of the joints.
It's not the ideal output but any other way we need to rethink the design of the setting up the optimization variables.

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:349]: This is by design. I did it this way so that if you specify a pram_joint that is not contained in the joints of the robot.
It neither gives an error nor does anything wrong.
It just filters out all the requested parametric joints that are not in the list of actuated joints.

File: optas/builder.py
Comment (cmower) [reply optas/builder.py:278]: Did I? Sorry, I think because it hadn't occurred to me before that this functionality could be abstracted to the `TaskModel` class. Keep this as it is and we can change later. As I mention elsewhere, keep this PR for the `RobotModel` and create an issue - we can discuss there these sorts of changes.

File: example/figure_eight_plan_6dof.py
Comment (cmower) [reply example/figure_eight_plan_6dof.py:112]: I don't think so. As above, I think what you implement here can be put in the `solve` method.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:95]: Done. Used the name num_opt_joints to be a bit more explicit.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:278]: Added new issue.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:312]: Done.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:307]: Done.

File: example/figure_eight_plan_6dof.py
Comment (joaomoura24) [reply example/figure_eight_plan_6dof.py:103]: Ok, i removed the zero initialization.

Comment (joaomoura24) [reply example/figure_eight_plan_6dof.py:103]: Done.

File: example/figure_eight_plan_6dof.py
Comment (joaomoura24) [reply example/figure_eight_plan_6dof.py:112]: As we saw this would require some major changes including changes to some of the nomenclature.
I would stay for know with this work around for this particular case and leave the changes for another PR.
The easier solution would be to include a function in the RobotModel for get_q instead of using the current solution[f'{robot_name}/q'].

File: example/figure_eight_plan_6dof.py
Comment (cmower) [reply example/figure_eight_plan_6dof.py]: Yes, but what you are specifying here does not impact the decision variables. You are making a constraint
```
g(p) = 0
```
rather than
```
g(x; p) = 0
```
If you want to put a non-zero configuration then this will not happen - this is because you are specifying the constraint incorrectly.

**Review by cmower — CHANGES_REQUESTED** [review]
Looking good, just some fairly minor points. I was thinking, could you double check the helper methods in the `OptimizationBuilder` class. I think some of these methods might break given this feature. E.g. `enforce_model_limits` may need to have the split/extract method called inside to make sure constraints are only enforced on optimized joints.

File: optas/solver.py
Comment (cmower) [inline optas/solver.py:114]: This can fail when `n_s_p` does not appear in `_p_dict`. The user can choose to not provide all the parameters, and in this case they are assumed to be zero. The way to get round this is as follows.
```python
solution[n_s][model.parameter_joint_indexes, :] = self._p_dict.get(n_s_p, default)
```
where `default = cs.DM.zeros(model.dim, model.T)`.

**NOTE**: this will need a small modification to the `OptimizationBuilder` class in the `__init__` method. I'll make a note there too.

File: optas/builder.py
Comment (cmower) [inline optas/builder.py:93]: This should be modified (this relates to a comment I made in the `_solve` method (see the one with **NOTE**). Please change to 
```python
for model in self._models:
    if model.T is None:
        model.T = T
    for d in model.time_derivs:
        n_s = model.state_name(d)
        t = model.T - d if not derivs_align else T
```

**Review by cmower — CHANGES_REQUESTED** [review]
Some more changes please.

File: optas/builder.py
Comment (cmower) [reply optas/builder.py:93]: This is actually not needed now (I realized in the above that the dimensions for the zeros array are readily available without making this change). However, I think it would be a good idea to make this change anyway. Makes the code a little cleaner. Also, model.T shouldn't be left as `None`.

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:78]: Done.

File: optas/solver.py
Comment (joaomoura24) [reply optas/solver.py:94]: Actually, i need to.
Because otherwise if the user does not initialize the parameter to zero, they won't be part of the _p_dict and then it will fail once we try to merge.
This way, even if the user does not initialize them the first line will do it and the second will just store it as a dictionary.
In summary, p is different from self._p_dict

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:391]: Done.

File: optas/models.py
Comment (joaomoura24) [reply optas/models.py:409]: Done.

File: optas/solver.py
Comment (joaomoura24) [reply optas/solver.py:111]: Done.

File: optas/solver.py
Comment (joaomoura24) [reply optas/solver.py:118]: Yes, but the solution is only with the name '/q/x'.
This line is adding the solution to the name '/q'.
Which for the case where there is no split then it is a repetition of solution, but in this way we can always collect the solution as '/q' which is cleaner.

File: optas/solver.py
Comment (joaomoura24) [reply optas/solver.py:112]: Done.

File: optas/solver.py
Comment (joaomoura24) [reply optas/solver.py:103]: Done

File: optas/solver.py
Comment (joaomoura24) [reply optas/solver.py:114]: I solved this problem.
Check my answer above. And the solution seems to me to be cleaner.

File: optas/builder.py
Comment (joaomoura24) [reply optas/builder.py:93]: Done.

File: optas/solver.py
Comment (cmower) [reply optas/solver.py:94]: Yes, but `vec2dict` assumes the input is a vector. The input here is a dictionary - this will fail. Handling the case when the user decides not to provide the parameters can be handled in `_solve`.

File: optas/solver.py
Comment (cmower) [reply optas/solver.py:119]: Ok, agreed

File: optas/solver.py
Comment (cmower) [reply optas/solver.py:115]: I think what you do above will fail. Perhaps make this change? And revert above.

File: optas/solver.py
Comment (joaomoura24) [reply optas/solver.py:94]: First, it won't fail, because I tested it. And it doesn't fail.

The p is a dictionary. The first dict2vec turns saves it as a self.p which is a vector. (but with all the parameters initialised.
The self.p is again a vector, so I just turn it again to a dictionary self._p_dict.

Maybe I can change the names to make it more clear.

I don't agree that the default parameters should be handled by the _solve(). This is a already quite cluttered. And as it is know the reset_parameters is handling both cases of of setting the default parameters.

File: optas/solver.py
Comment (joaomoura24) [reply optas/solver.py:115]: No, it doesn't fail.
The reset_parameters() makes sure that they are initialised as you were doing before.
It is just doing the same thing as before.

File: optas/solver.py
Comment (cmower) [reply optas/solver.py:94]: Aaahh, sorry. I was reading this as if it were
```python
self._p_dict = self.opt.parameters.vec2dict(p)
```
which would fail. Yes, you're correct - this would pass fine as you have it.

Comment (cmower) [reply optas/solver.py:94]: Ok, all good!

File: optas/solver.py
Comment (cmower) [reply optas/solver.py:115]: Ok, all good.

**Comment by joaomoura24** [comment]
Don't merge this yet.
I am fixing the examples to run with the new 'q/x'.

**Comment by joaomoura24** [comment]
I updated the examples and ran them.
Note that i got errors in the examples to do with the issues #77 and #78.
I ran those two examples in the master branch as well just to check if the problem was still there.

**Comment by joaomoura24** [comment]
I think this is ready to merge.
It still has the Changes requested on for some reason.
Could you take a look at it and see if you can merge?

**Comment by cmower** [comment]
Thanks @joaomoura24!

**Pull request merged**
